persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/JdbcBasePersistenceImpl.java

```diff
@@ -459,29 +461,37 @@ public <T> Page<T> listEntities(
             entityType.getCode(),
             "realm_id",
             realmId);
+    Map<String, Object> whereGreater = Map.of();
 
     // Limit can't be pushed down, due to client side filtering
     // absence of transaction.
+    String orderByColumnName = null;
+    if (pageToken.paginationRequested()) {
+      orderByColumnName = ModelEntity.ID_COLUMN;
+      if (pageToken.hasDataReference()) {
+        long boundary = entityIdBoundary(pageToken);
+        whereGreater = Map.of(ModelEntity.ID_COLUMN, boundary);
+      }
+    }
+
     try {
       PreparedQuery query =
           QueryGenerator.generateSelectQuery(
-              ModelEntity.ALL_COLUMNS, ModelEntity.TABLE_NAME, params);
-      List<PolarisBaseEntity> results = new ArrayList<>();
+              ModelEntity.ALL_COLUMNS,
+              ModelEntity.TABLE_NAME,
+              whereEquals,
+              whereGreater,
+              orderByColumnName);
+      AtomicReference<Page<T>> results = new AtomicReference<>();
       datasourceOperations.executeSelectOverStream(
           query,
           new ModelEntity(),
           stream -> {
             var data = stream.filter(entityFilter);
-            if (pageToken instanceof HasPageSize hasPageSize) {
-              data = data.limit(hasPageSize.getPageSize());
-            }
-            data.forEach(results::add);
+            results.set(
+                Page.mapped(pageToken, data, transformer, EntityIdPaging::encodedDataReference));
```

Comment:
Why are we pushing the limit into `Page.mapped`? This feels beyond the scope of what a `Page` should be handling. If we _are_ pushing it down, why not also push the filter?

Reply:
The filter (line 490) is implicitly part of the page construction login by being a stage in the `data` stream.

The use of `PageToken.pageSize()` inside `Page.mapped()` is merely a convenience / guardrail to avoid assuming that the stream is limited to the exact page size.